### ISSUE TITLE
Update CI for `setup`?

### ISSUE THREAD

**Issue by andreasabel** [haskell/actions#116]
CI for `setup` focuses of GHC 8.4.4, and gives special mention to 7.4.1, 7.10.3, 8.2.2, see https://github.com/haskell/actions/actions/runs/3156574295.

`latest` is also covered, but shouldn't extra testing go to more recent GHC versions?

I would be good if the `workflow.yml` file contained some documentation of the testing policy, such that it can be kept up to date also by external contributors (like me).  Atm, I don't have a clear idea what CI aims to achieve (in terms of covering GHC/cabal versions), so I cannot sensibly make modifications.

**Comment by hazelweakly** [comment]
This is a good point :)

The CI is supposed to test a few things but doesn't particularly call them out super well

Essentially, I needed to test the fallback methods of the tools as they fail various edge cases and go back to other ones. I also needed to test old as well as new packages, and weird version resolution edge cases.

I annotated a bunch of these with details and updated versions here and there. I _think_ it's an improvement?

Part of the reason testing isn't done on most recent versions is that, generally speaking, if `latest` works, the last 3 will work fine, because so much of the heavy lifting of this action is actually done by `ghcup` or `chocolatey`.

https://github.com/haskell/actions/tree/update-ghc-and-ci

This branch contains an annotated CI file that hopefully is much clearer in intent
